### PR TITLE
v1.0.0

### PR DESCRIPTION
**Pull request by alfonsomunozpomer**
A big re-engineering of the whole thing, so here are some design principles I followed:

## Cards
Cards are very dumb and are the components meant to be reusable.

They share `cardPropTypes` which is an implementation of the JSON schema as props, plus narrowing `iconType` to two values (@monicajianu is that possible to express in the schema?).

Styles are applied sparingly with regular objects, although we could go with `styled-components` so that instead of inline styles they get a scoped class. Any thoughts?

To have feature equivalence between both types of cards, `SpeciesCard` could be something like `StaticCard` and be able to render both `species` and `image`. As it stands `ExtendableCard` is more generic and they are semantically a bit different, but that’s ok for now.

I’ve refactored the list items but it may be my being overzealous about drying up the code.

## Containers
They mix and match cards as they want, and are tightly coupled to our back end. Here I’ve used `styled-components` to properly style `hover`. We’ll be able to remove a couple of styles in `atlas.css` too. 

I’ve made FetchLoader as a HOC (*finally!*). This is a big step in putting it in a separate repository and compose it when necessary.

Also, an image is worth more than a thousand words, so here is the new demo page (the HCA container is missing since I need to review the endpoint):

image

### CODE REVIEW

File: src/containers/SceaHomepageSpeciesContainer.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import SpeciesCard from '../cards/SpeciesCard'
+
+const CardContainer = styled.div`
+  border-radius: 8px;
+  :hover {
+    background: AliceBlue;
+  }
+`
+
+const SceaHomepageSpeciesContainer = ({cards}) =>
```
Comment (lingyun1010) [inline src/containers/SceaHomepageSpeciesContainer.js:15]: Is it a typo here, `Scea` not `Scxa`?

File: src/cards/ExtendableCard.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        <ul className={`content`} style={{listStyle: `none`}}>
+        {
+          this.state.isHidden ?
+            visibleContent.slice(0, MAX) :
+            visibleContent
+        }
+        </ul>
+        {
+          Array.isArray(content) && content.length > MAX &&
+          <button className={`button`} onClick={this.onClick}>{this.state.isHidden ? `Show all` : `Show less`}</button>
```
Comment (monicajianu) [inline src/cards/ExtendableCard.js]: Show fewer, not less.

File: src/containers/HcaLandingPageContainer.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import styled from 'styled-components'
+
+import withFetchLoader from './FetchLoader'
```
Comment (monicajianu) [inline src/containers/HcaLandingPageContainer.js]: Unused import

File: src/containers/SceaHomepageSpeciesContainer.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import styled from 'styled-components'
+
+import withFetchLoader from './FetchLoader'
```
Comment (monicajianu) [inline src/containers/SceaHomepageSpeciesContainer.js]: Unused import

File: src/containers/SceaHomepageSpeciesContainer.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import cardPropTypes from '../cards/cardPropTypes'
+import SpeciesCard from '../cards/SpeciesCard'
+
+const CardContainer = styled.div`
+  border-radius: 8px;
```
Comment (monicajianu) [inline src/containers/SceaHomepageSpeciesContainer.js:9]: Hmm I got rid of this rounded corner in another commit (https://github.com/ebi-gene-expression-group/atlas/pull/185/commits/ecc1bad138e0fc4fd9aed4f9e427c74aeb9fbce9). It's literally the only place in the entire application where we use rounded corners. Not very consistent. I also changed the blue to be more EBI-visual-framework-friendly. We don't want to many shades of blue floating around.

Comment (monicajianu) [reply src/containers/SceaHomepageSpeciesContainer.js:9]: Same comments apply to `HcaLandingPageContainer`

File: src/containers/HcaLandingPageContainer.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+import ExtendableCard from '../cards/ExtendableCard'
+
+const CardContainer = styled.div`
+  border-radius: 8px;
+  :hover {
+    background: AliceBlue;
+  }
+`
+
+const HcaCardContainer = ({cards}) =>
```
Comment (monicajianu) [inline src/containers/HcaLandingPageContainer.js]: This component could be renamed to be more generic. There is nothing specifically HCA about it. At the very least it should match the name of the `.js` file.

File: src/cards/SpeciesCard.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        description.url ?
+          <a href={description.url}>{description.text}</a> :
+          description.text
+      }
+      </h5>
+    }
+
+    {
+      Array.isArray(content) &&
+      <ul style={{listStyle: `none`}}>
```
Comment (monicajianu) [inline src/cards/SpeciesCard.js]: Set `margin-left: 0` to get rid of the extra space and centre the text properly.

Comment (monicajianu) [reply src/cards/SpeciesCard.js]: Same comment applies to `ExtendableCard`

File: src/cards/ExtendableCard.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    })
+  }
+
+  render() {
+    const { iconType, iconSrc, description, content } = this.props
+    const visibleContent = Array.isArray(content) && renderContentListItems(content)
+
+    const cardHero =
+      iconType === `species` ?
+        <span style={{fontSize: `8rem`}}><EbiSpeciesIcon species={iconSrc} /></span> :
```
Comment (monicajianu) [inline src/cards/ExtendableCard.js:33]: Font size is `rem` here but `%` in the other card

File: src/cards/ExtendableCard.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
+import React from 'react'
+import PropTypes from 'prop-types'
```
Comment (monicajianu) [inline src/cards/ExtendableCard.js:2]: Unused import

File: src/cards/SpeciesCard.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
+import React from 'react'
+import PropTypes from 'prop-types'
```
Comment (monicajianu) [inline src/cards/SpeciesCard.js:2]: Unused import

File: src/cards/ExtendableCard.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+        <ul className={`content`} style={{listStyle: `none`}}>
+        {
+          this.state.isHidden ?
+            visibleContent.slice(0, MAX) :
+            visibleContent
+        }
+        </ul>
+        {
+          Array.isArray(content) && content.length > MAX &&
```
Comment (monicajianu) [inline src/cards/ExtendableCard.js:63]: This is a bit verbose... Could we have a variable `hasExtendableContent`/`hasMoreContent`/`insert-more-inspired-name-here = Array.isArray(content) && content.length > MAX`? Just to make this conditional rendering a bit more explicit.

File: src/cards/ExtendableCard.js [hunk 12]
```diff
@@ -0,0 +1,73 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+
+import EbiSpeciesIcon from 'react-ebi-species'
+
+import renderContentListItems from './renderContentListItems'
+import cardPropTypes from './cardPropTypes'
+
+const MAX = 5
```
Comment (monicajianu) [inline src/cards/ExtendableCard.js:9]: `MAX` is a bit on the generic side. It's making me want to ask "max what"?

File: src/cards/ExtendableCard.js [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+        {
+          description && (
+            description.url ?
+              <a href={description.url} style={{borderBottom: 0}}>{cardHero}</a> :
+              cardHero
+            )
+        }
+
+        <ul className={`content`} style={{listStyle: `none`}}>
```
Comment (monicajianu) [inline src/cards/ExtendableCard.js]: This `<ul>` element will be rendered even if there is no content...

File: src/containers/SceaHomepageSpeciesContainer.js (hunk 1 above)
Comment (monicajianu) [inline src/containers/SceaHomepageSpeciesContainer.js:15]: Similar comment as before, this could have a more generic name. It's basically a grid of species cards, regardless of where it is used in SC.

File: src/index.js [hunk 14]
```diff
@@ -1,3 +1,11 @@
-import CardContainer from './CardContainer.js'
+import withFetchLoader from './containers/FetchLoader'
+import SceaHomepageSpeciesContainer from './containers/SceaHomepageSpeciesContainer'
+import HcaLandingPageContainer from './containers/HcaLandingPageContainer'
 
-export default CardContainer
+const _SceaHomepageSpeciesContainer = withFetchLoader(SceaHomepageSpeciesContainer)
```
Comment (monicajianu) [inline src/index.js:5]: 👍 This HOC stuff really is magical!

File: html/index.html [hunk 15]
```diff
@@ -1,60 +1,203 @@
...
-    resource: 'json/experiments/popular-species'
-  }, 'target')
+  homepageCardsDemo.renderExtendableCard({
+    iconType: `image`,
+    iconSrc: `https://upload.wikimedia.org/wikipedia/en/6/6b/A_Rickle_in_Time.jpg`,
+    description: {
+      text: `A rickle in time`,
+      url: `http://rickandmorty.wikia.com/wiki/A_Rickle_in_Time`
+    },
+    content: [
```
Comment (monicajianu) [inline html/index.html:54]: 🤓!!! I'm clearly missing out on this whole Rick and Morty universe...

File: __test__/SpeciesCard.test.js [hunk 16]
```diff
@@ -1,74 +1,55 @@
...
 
-  test(`does not render optional empty content`, () => {
-    const wrapper = shallow(<SpeciesCard iconSrc={``} />)
+  test(`does not render optional empty title and content`, () => {
+    const wrapper = shallow(<SpeciesCard iconType={`species`} iconSrc={``} />)
 
     expect(wrapper.find(EbiSpeciesIcon).exists()).toBe(true)
-    expect(wrapper.find(`.species-name`).exists()).toBe(false)
-    expect(wrapper.find(`.content`).exists()).toBe(false)
+    expect(wrapper.find(`h5`).exists()).toBe(false)
```
Comment (monicajianu) [inline __test__/SpeciesCard.test.js:26]: Hmmm. Not sure about this. I mean it works in this small context, but it's still not clear that `h5` is the title and `ul` and `li` are part of the content... (unless you're intimately familiar with the code) Also, if you were doing testing at a more end-to-end, Selenium level, having those meaningful CSS class names really helps!

File: __test__/TestUtils.js [hunk 17]
```diff
@@ -1,75 +1,157 @@
-
-
 // Stolen from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random
```
Comment (monicajianu) [inline __test__/TestUtils.js:1]: I probably didn't mean this exact wording to make it into the final thing... 😟  A more professional sounding attribution might be needed... 😊

**Review by monicajianu — CHANGES_REQUESTED** [review]
Just a couple of minor things here and there.

File: __test__/TestUtils.js (hunk 17 above)
Comment (alfonsomunozpomer) [reply __test__/TestUtils.js:1]: IIRC I used that wording elsewhere and it’s been copy-pasted around.

File: src/containers/SceaHomepageSpeciesContainer.js (hunk 1 above)
Comment (alfonsomunozpomer) [reply src/containers/SceaHomepageSpeciesContainer.js:15]: So, history time!
`gxa` was the moniker given to Expression Atlas, as the original project name was “Gene Expression Atlas”, and the “x” was used instead of “e” to give the URL more personality and make it easier to remember. However, nowadays, we prefer to use EA when using an acronym to refer to Expression Atlas.
`scxa` was the internal name of Single Cell Expression Atlas to follow the pattern of `gxa`, but again, we use SCEA more often.

In summary: any way is right, but consider `gxa` and `scxa` historical curiosities.

File: src/cards/ExtendableCard.js (hunk 2 above)
Comment (alfonsomunozpomer) [reply src/cards/ExtendableCard.js]: Tbh, I think less is also grammatically correct, but it’s probably abusing the term.

File: src/containers/SceaHomepageSpeciesContainer.js (hunk 5 above)
Comment (alfonsomunozpomer) [reply src/containers/SceaHomepageSpeciesContainer.js:9]: We are using these styles after Seb’s work when we migrated the site to the EBI VF. I don’t really think the choice is more or less EBI VF-friendly, what do you mean exactly? That it’s closer to the original colours? One of the things the framework explicitly states is that you can choose a theme for your project.

We went with a kind of sky blue as the primary colour for buttons, and `AliceBlue` to highlight the items in the home page and also all tabs (front page and experiment page). I think the colours are used judiciously.

Comment (monicajianu) [reply src/containers/SceaHomepageSpeciesContainer.js:9]: Point taken for the blue, I didn't know the story behind it. But I still stand behind my rounded corners comment.

File: __test__/TestUtils.js (hunk 17 above)
Comment (monicajianu) [reply __test__/TestUtils.js:1]: Ahh good, it just sounded like something I would write, though I didn't remember writing it... Explains why!

File: html/index.html [hunk 18]
```diff
@@ -45,69 +45,69 @@ <h2 class="text-center">Image card container</h1>
 <!-- Set to http://localhost:8080/gxa/ or http://localhost:8080/gxa/sc/ -- Remember the trailing slash! -->
 <script>
   homepageCardsDemo.renderExtendableCard({
-    iconType: `image`,
-    iconSrc: `https://upload.wikimedia.org/wikipedia/en/6/6b/A_Rickle_in_Time.jpg`,
+    iconType: 'image',
```
Comment (monicajianu) [inline html/index.html:48]: Oh? I thought it was meant to be backticks everywhere?

**Review by monicajianu — APPROVED** [review]

File: html/index.html (hunk 18 above)
Comment (alfonsomunozpomer) [reply html/index.html:48]: Most moder browsers can interpret them correctly, but I wouldn’t bet on that. Notice the semicolons, too.
